File: android/core/src/main/kotlin/com/openmausbot/companion/core/Onboarding.kt
package com.openmausbot.companion.core

/**
 * The first-run decision seam — the port of
 * `ios/Sources/CompanionCore/Onboarding.swift`.
 *
 * It lives outside Compose for the same reason it lives outside SwiftUI: the
 * transitions that matter are the ones nobody sees until they are wrong.
 * Skipping setup must not look like a pairing, a deep link that arrived while
 * the app was closed must still open pairing, and a revoked credential must
 * never fall through to an ordinary empty state.
 *
 * What is ported here is the **timing and the routing**. The painting is
 * Material's, and deliberately not SwiftUI's.
 */
enum class OnboardingPairingState {
    UNPAIRED,
    PAIRED,
    REVOKED,
}

enum class OnboardingRoute {
    WELCOME,
    PAIRING,
    UNPAIRED_HOME,
    NOTIFICATION_PROMPT,
    CHATS,
    REVOKED,
}

/**
 * Whether the app yet knows what the system will say about notifications.
 *
 * iOS reads this asynchronously at launch, and treating that short unresolved
 * window as a final answer can skip first-pair education forever. Android
 * answers synchronously today, so [UNRESOLVED] is produced by the durable side
 * instead — see `OnboardingPreferences` in `:app`. The rule the third case
 * carries is not iOS's async quirk: **nothing may spend the marker on an
 * answer the app does not have.**
 */
enum class NotificationAuthorizationState {
    UNRESOLVED,
    NOT_DETERMINED,
    DETERMINED,
}

/**
 * Durable preference names shared by the pairing commit and the root router.
 *
 * The pending marker is deliberately separate from the benign "already saw
 * this" preference: a new pairing may finish before the app knows the current
 * notification authorization, and the two answers are not the same question.
 *
 * None of these is a secret and none may become a place where anything else
 * rides along (§6): the store that holds them accepts booleans and nothing
 * else, and `OnboardingPreferencesContractTest` in `:app` holds it to that.
 */
object OnboardingPreferenceKeys {
    const val PENDING_NOTIFICATION_ONBOARDING = "companion.onboarding.notificationPending"
    const val WELCOME_SEEN = "companion.onboarding.welcomeSeen"
    const val NOTIFICATION_PROMPT_SEEN = "companion.onboarding.notificationsSeen"

    /** Everything this app is allowed to keep in the onboarding store. */
    val ALL: Set<String> = setOf(
        PENDING_NOTIFICATION_ONBOARDING,
        WELCOME_SEEN,
        NOTIFICATION_PROMPT_SEEN,
    )
}

/**
 * Keeps the crash-sensitive half of a successful pairing commit explicit, and
 * therefore testable.
 *
 * The notification marker has to exist **before** the connection becomes
 * restorable. An orphan marker while unpaired is harmless — the router only
 * reaches the education step from [OnboardingPairingState.PAIRED] — but a
 * connection restored without its marker permanently skips first-pair
 * education, and there is no later moment that can tell it should not have.
 */
object PairingCommitSequence {
    suspend fun persist(
        markNotificationOnboardingPending: suspend () -> Unit,
        saveConnection: suspend () -> Unit,
    ) {
        markNotificationOnboardingPending()
        saveConnection()
    }
}

/** Pure lifecycle policy for the durable first-pair notification marker. */
object NotificationOnboardingPolicy {
    fun shouldKeepPending(
        isPending: Boolean,
        hasCompletedStep: Boolean,
        authorization: NotificationAuthorizationState,
    ): Boolean {
        if (!isPending) return false
        // A relaunch can restore the pairing before the app knows what the
        // system will say. Never spend the marker during that window.
        if (authorization == NotificationAuthorizationState.UNRESOLVED) return true
        // Once the answer is known, education is needed only while the system
        // can still be asked and this person has not already done or skipped
        // the step.
        return authorization == NotificationAuthorizationState.NOT_DETERMINED && !hasCompletedStep
    }
}

data class OnboardingContext(
    val pairingState: OnboardingPairingState,
    val hasSeenWelcome: Boolean,
    val pairingRequested: Boolean = false,
    val hasPendingPairingInvite: Boolean = false,
    /**
     * Written only when a new pairing commits. People who were already paired
     * do not receive first-pair education merely because they updated the app.
     */
    val notificationOnboardingPending: Boolean = false,
    val hasSeenNotificationPrompt: Boolean = false,
    val notificationAuthorization: NotificationAuthorizationState =
        NotificationAuthorizationState.NOT_DETERMINED,
)

object OnboardingRouter {
    fun route(context: OnboardingContext): OnboardingRoute = when (context.pairingState) {
        // First, and unconditionally. A revoked token with a pending invite and
        // a requested pairing is still a revoked token: recovery outranks every
        // other reason this screen could be shown.
        OnboardingPairingState.REVOKED -> OnboardingRoute.REVOKED

        OnboardingPairingState.PAIRED ->
            if (context.pairingRequested || context.hasPendingPairingInvite) {
                // Adding a computer deliberately overlays pairing over a live
                // session. The old computer is not replaced until commit.
                OnboardingRoute.PAIRING
            } else if (context.notificationOnboardingPending &&
                !context.hasSeenNotificationPrompt &&
                context.notificationAuthorization == NotificationAuthorizationState.NOT_DETERMINED
            ) {
                OnboardingRoute.NOTIFICATION_PROMPT
            } else {
                OnboardingRoute.CHATS
            }

        OnboardingPairingState.UNPAIRED ->
            if (context.pairingRequested || context.hasPendingPairingInvite) {
                OnboardingRoute.PAIRING
            } else if (context.hasSeenWelcome) {
                OnboardingRoute.UNPAIRED_HOME
            } else {
                OnboardingRoute.WELCOME
            }
    }
}

File: ios/Sources/CompanionCore/Onboarding.swift
/// A small, platform-neutral decision seam for the companion's first-run flow.
///
/// Keeping this outside SwiftUI makes the important transitions explicit and
/// testable: skipping setup must not look like a pairing, a pending deep link
/// must still open pairing, and a revoked credential must never fall through
/// to an ordinary empty state.
public enum CompanionPairingState: Equatable, Sendable {
    case unpaired
    case paired
    case revoked
}

public enum CompanionOnboardingRoute: Equatable, Sendable {
    case welcome
    case pairing
    case unpairedHome
    case notificationPrompt
    case chats
    case revoked
}

/// The permission lookup is asynchronous at launch. Treating that short
/// unresolved window as a final answer can skip first-pair education forever.
public enum CompanionNotificationAuthorizationState: Equatable, Sendable {
    case unresolved
    case notDetermined
    case determined
}

/// Durable preference names shared by the app's pairing commit and its root
/// router. The pending marker is intentionally separate from the benign
/// "already saw this" preference: a new pairing may finish before iOS returns
/// the current notification authorization status.
public enum CompanionOnboardingPreferences {
    public static let pendingNotificationOnboardingKey =
        "companion.onboarding.notificationPending"
}

/// Keeps the crash-sensitive part of a successful pairing commit explicit
/// and testable. The notification marker must exist before the restorable
/// connection: an orphan marker while unpaired is harmless, but a connection
/// without its marker can permanently skip first-pair education.
public enum CompanionPairingCommitSequence {
    public static func persist(
        markNotificationOnboardingPending: () -> Void,
        saveConnection: () -> Void
    ) {
        markNotificationOnboardingPending()
        saveConnection()
    }
}

public enum CompanionPairingInviteEvent: Equatable, Sendable {
    case received(PairingInvite)
    case consumed
    case pairingSucceeded
    case signedOut
}

/// Pure invite lifecycle shared by Session and sequence tests. A paired phone
/// may receive another computer's invite; PairingView keeps the current
/// connection alive until the new credential is safely committed.
public enum CompanionPairingInvitePolicy {
    public static func nextInvite(
        current: PairingInvite?,
        after event: CompanionPairingInviteEvent
    ) -> PairingInvite? {
        switch event {
        case .received(let invite):
            return invite
        case .consumed, .pairingSucceeded, .signedOut:
            return nil
        }
    }
}

/// Pure lifecycle policy for the durable first-pair notification marker.
public enum CompanionNotificationOnboardingPolicy {
    public static func shouldKeepPending(
        isPending: Bool,
        hasCompletedStep: Bool,
        authorization: CompanionNotificationAuthorizationState
    ) -> Bool {
        guard isPending else { return false }
        // A relaunch can restore the pairing before UserNotifications has
        // answered. Never spend the marker during that temporary state.
        guard authorization != .unresolved else { return true }
        // Once status is known, education is needed only when iOS can still
        // ask and this user has not already completed or skipped the step.
        return authorization == .notDetermined && !hasCompletedStep
    }
}

/// A small state machine used by PairingView to make navigation mutually
/// exclusive with a pairing commit. A second submit or reset cannot overtake
/// the request which may already have persisted on the Mac and phone.
public struct CompanionPairingSubmissionState: Equatable, Sendable {
    public private(set) var isInFlight = false

    public init() {}

    public var allowsNavigation: Bool { !isInFlight }

    @discardableResult
    public mutating func begin() -> Bool {
        guard !isInFlight else { return false }
        isInFlight = true
        return true
    }

    public mutating func finish() {
        isInFlight = false
    }
}

public struct CompanionOnboardingContext: Equatable, Sendable {
    public var pairingState: CompanionPairingState
    public var hasSeenWelcome: Bool
    public var pairingRequested: Bool
    public var hasPendingPairingInvite: Bool
    /// Persisted only after a new pairing commits. Existing paired users do
    /// not receive first-pair education merely because they upgraded.
    public var notificationOnboardingPending: Bool
    public var hasSeenNotificationPrompt: Bool
    public var notificationAuthorization: CompanionNotificationAuthorizationState

    public init(
        pairingState: CompanionPairingState,
        hasSeenWelcome: Bool,
        pairingRequested: Bool = false,
        hasPendingPairingInvite: Bool = false,
        notificationOnboardingPending: Bool = false,
        hasSeenNotificationPrompt: Bool = false,
        notificationAuthorization: CompanionNotificationAuthorizationState = .notDetermined
    ) {
        self.pairingState = pairingState
        self.hasSeenWelcome = hasSeenWelcome
        self.pairingRequested = pairingRequested
        self.hasPendingPairingInvite = hasPendingPairingInvite
        self.notificationOnboardingPending = notificationOnboardingPending
        self.hasSeenNotificationPrompt = hasSeenNotificationPrompt
        self.notificationAuthorization = notificationAuthorization
    }
}

public enum CompanionOnboardingRouter {
    public static func route(for context: CompanionOnboardingContext) -> CompanionOnboardingRoute {
        switch context.pairingState {
        case .revoked:
            return .revoked
        case .paired:
            if context.pairingRequested || context.hasPendingPairingInvite {
                return .pairing
            }
            if context.notificationOnboardingPending,
               !context.hasSeenNotificationPrompt,
               context.notificationAuthorization == .notDetermined {
                return .notificationPrompt
            }
            return .chats
        case .unpaired:
            if context.pairingRequested || context.hasPendingPairingInvite {
                return .pairing
            }
            return context.hasSeenWelcome ? .unpairedHome : .welcome
        }
    }
}
